### PR TITLE
Store decode notes

### PR DESCRIPTION
**Pull request by panleone**
Abstract:
This PR will add the feature of storing, decoding  and checking if a note is spent
Also allow to create the Merkle tree from the closest checkpoint

### CODE REVIEW

File: src/checkpoint.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+//Return an array of satic ordered checkpoints
+fn get_checkpoint(block_height: i32, is_testnet: bool) -> (i32, &'static str) {
```
Comment (Duddino) [inline src/checkpoint.rs:2]:
```suggestion
fn get_checkpoint(block_height: i32, is_testnet: bool) -> Option<(i32, &'static str)> {
```
Change this to an option so we don't have to panic if there is no available checkpoint

File: src/checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    let test_network_checkpoins = [(1125777, "018c325f63f5cc98541cfef957f64845c86cf928e317ecc71a14debd364c7b8f57013c6f50deb5f788d5ac9105915ab9cbcda21a101d267c6424aa75b6e8df969e480d00016a2b0e3728a820b7982d81c87b80468ce65a4081843b890307115ca896416f3901e105bf42db29eca36e7235bd55546726753d1f967c3f284e243cbb3b3375d95a01a3ce8339e68a22d91b0750ef45468efe763e3d5e3a6e59809ddadcc94fe73c6c01494803bd8e6b730cb277701c613a4e7355cb54f79653724618e1436c02fca30c0177d25b5ed812af45eb46b54bc37c3fbe08fdfb4d952bb917fe59187bc78c42640001088b8a9fc4769017f3fdf865637e5cebbeaf7a4c643247723bf009da5eb1e4340001e877753448933a336fcf9399cc3dcd357344510c79db717e976979cb2eab612d0001cb846820acd916b4ea03b0a222b3eae8704bbd5365f105156041c578bd214c3201e03719b3810c7a9eaf6680ad3c60fb5ffdb0106975c952ef173c3e8cde943b03")];
+    let main_network_checkpoins = [(0, "0")];
+    let used_checkpoints = if is_testnet {
+        test_network_checkpoins
+    } else {
+        main_network_checkpoins
+    };
+    return used_checkpoints.iter().rev().filter(|x| { x.0 < block_height}).next().unwrap().clone();
```
Comment (Duddino) [inline src/checkpoint.rs:10]:
```suggestion
    return used_checkpoints.iter().rev().filter(|x| { x.0 < block_height}).next().clone();
```

File: src/lib.rs [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
 use pivx_primitives::zip32::sapling::ExtendedSpendingKey;
 use pivx_primitives::zip32::AccountId;
 use pivx_primitives::zip32::DiversifierIndex;
 
+use pivx_client_backend::decrypt_transaction;
 use pivx_client_backend::encoding;
-use pivx_client_backend::keys::sapling; //::{decode_extended_spending_key, decode_transparent_address};
-                                        // When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
-                                        // allocator.
+use pivx_client_backend::keys::{sapling, UnifiedFullViewingKey}; //::{decode_extended_spending_key, decode_transparent_address};
```
Comment (Duddino) [inline src/lib.rs:23]: Remove the comment

File: src/lib.rs [hunk 4]
```diff
@@ -90,6 +100,60 @@ pub fn generate_next_shielding_payment_address(
     }
 }
 
+fn handle_block(
+    tree: &mut CommitmentTree<Node>,
+    block_data: String,
```
Comment (Duddino) [inline src/lib.rs:105]:
```suggestion
    block_data: &str,
```
Change to &str so we don't have to use the owned data type

File: src/lib.rs [hunk 5]
```diff
@@ -90,6 +100,60 @@ pub fn generate_next_shielding_payment_address(
...
+    key: &UnifiedFullViewingKey,
+    isTestnet: bool,
+) -> Vec<(Note, MerklePath<Node>)> {
```
Comment (Duddino) [inline src/lib.rs:108]:
```suggestion
) -> Result<Vec<(Note, MerklePath<Node>)>, Box<dyn Error>> {
```

**Review by Duddino — CHANGES_REQUESTED** [review]

**Comment by panleone** [comment]
Ready for review

File: src/lib.rs [hunk 6]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+    let mut tree = CommitmentTree::<Node>::read(buff).expect("Cannot decode commitment tree!");
+    let extsk = decode_extsk(&enc_extsk, is_testnet);
+    let key = UnifiedFullViewingKey::new(Some(extsk.to_diversifiable_full_viewing_key()), None)
+        .expect("Failed to create unified full viewing key");
+    let (nullifiers, comp_note) =
+        handle_transaction_internal(&mut tree, &tx, &key, true).expect("Cannot decode tx");
+    let mut ser_comp_note: Vec<(Note, String)> = vec![];
+    let mut ser_nullifiers: Vec<String> = vec![];
+    for x in comp_note.iter() {
+        let mut buff = Cursor::new(Vec::new());
```
Comment (Duddino) [inline src/lib.rs:144]:
```suggestion
        let mut buff = Vec::new();
```
Cursor is not needed

File: src/lib.rs [hunk 7]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+        Cursor::new(hex::decode(tree_hex).expect("Cannot decode commitment tree from hexadecimal"));
+    let mut tree = CommitmentTree::<Node>::read(buff).expect("Cannot decode commitment tree!");
+    let extsk = decode_extsk(&enc_extsk, is_testnet);
+    let key = UnifiedFullViewingKey::new(Some(extsk.to_diversifiable_full_viewing_key()), None)
+        .expect("Failed to create unified full viewing key");
+    let (nullifiers, comp_note) =
+        handle_transaction_internal(&mut tree, &tx, &key, true).expect("Cannot decode tx");
+    let mut ser_comp_note: Vec<(Note, String)> = vec![];
+    let mut ser_nullifiers: Vec<String> = vec![];
+    for x in comp_note.iter() {
```
Comment (Duddino) [inline src/lib.rs:143]:
```suggestion
    for (note, witness) in comp_note.iter() {
```

File: src/lib.rs [hunk 8]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+        let mut buff = Cursor::new(Vec::new());
+        x.1.write(&mut buff)
+            .expect("Cannot write witness to buffer");
+        let newNote = x.0.clone();
```
Comment (Duddino) [inline src/lib.rs:147]:
```suggestion

```

File: src/lib.rs [hunk 9]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+        ser_comp_note.push((newNote, hex::encode(&buff.into_inner())));
```
Comment (Duddino) [inline src/lib.rs:148]:
```suggestion
        ser_comp_note.push((note.clone(), hex::encode(&buff.into_inner())));
```

File: src/lib.rs [hunk 10]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+    }
+
+    for nullif in nullifiers.iter() {
+        ser_nullifiers.push(hex::encode(nullif.0));
+    }
+
+    let mut buff = Cursor::new(Vec::new());
```
Comment (Duddino) [inline src/lib.rs:155]:
```suggestion
    let mut buff = Vec::new();
```

File: src/lib.rs [hunk 11]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+                    witnesses.push((note.note.clone(), witness));
+                }
+            }
+        }
+    }
+    Ok((nullifiers, witnesses))
+}
+
+#[wasm_bindgen]
+pub fn remove_unspent_notes(
```
Comment (Duddino) [inline src/lib.rs:208]:
```suggestion
pub fn remove_spent_notes(
```
Typo

File: src/lib.rs [hunk 12]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+        serde_wasm_bindgen::from_value(nullifiers_data).expect("Cannot deserialize nullifiers");
+    let mut notes: Vec<(Note, String, MerklePath<Node>)> = vec![];
+    let mut unspent_notes: Vec<(Note, String)> = vec![];
+
+    let extsk = decode_extsk(&enc_extsk, is_testnet);
+    let nullif_key = extsk
+        .to_diversifiable_full_viewing_key()
+        .to_nk(Scope::External);
+
+    for x in hex_notes.iter() {
```
Comment (Duddino) [inline src/lib.rs:226]: Destructor this same as above

File: src/lib.rs [hunk 13]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
+        let buff = Cursor::new(hex::decode(&x.1).expect("Cannot decode witness"));
+        let newNote = x.0.clone();
+        let path = IncrementalWitness::<Node>::read(buff)
+            .expect("Cannot read witness from buffer")
+            .path()
+            .expect("Cannot find witness path");
+        notes.push((newNote, x.1.clone(), path));
+    }
+    for x in notes.iter() {
```
Comment (Duddino) [inline src/lib.rs:235]: Destructure this

**Review by Duddino — CHANGES_REQUESTED** [review]
I like the logic, but there are some styling issues.
Also run cargo fmt

File: src/lib.rs [hunk 14]
```diff
@@ -74,23 +106,198 @@ pub fn generate_next_shielding_payment_address(
...
             }
         }
         diversifier_index.increment();
     }
 }
 
+//Output the closest checkpoint to a given blockheight
+#[wasm_bindgen]
+pub fn get_closest_checkpoint(block_height: i32, is_testnet: bool) -> JsValue {
+    return serde_wasm_bindgen::to_value(&get_checkpoint(block_height, is_testnet).unwrap())
```
[the remark this answers is not in the document]
Comment (panleone) [reply src/lib.rs:121]: I will fix this on the Checkpoint PR

**Comment by panleone** [comment]
Suggestion should be fixed now

**Review by Duddino — APPROVED** [review]
tACK

**Pull request merged**